pkg/api/endpoints/util.go

```diff
@@ -83,21 +65,63 @@ func RepackSubsets(subsets []api.EndpointSubset) []api.EndpointSubset {
 	// Next, build the N-to-M association the API wants.
 	final := []api.EndpointSubset{}
 	for key, ports := range addrSetsToPorts {
-		addrs := []api.EndpointAddress{}
-		for k := range addrSets[key] {
-			addrs = append(addrs, *k)
+		var addrs, notReadyAddrs []api.EndpointAddress
+		for k, ready := range addrSets[key] {
+			if ready {
+				addrs = append(addrs, *k)
+			} else {
+				notReadyAddrs = append(notReadyAddrs, *k)
+			}
 		}
-		final = append(final, api.EndpointSubset{Addresses: addrs, Ports: ports})
+		final = append(final, api.EndpointSubset{Addresses: addrs, NotReadyAddresses: notReadyAddrs, Ports: ports})
 	}
 
 	// Finally, sort it.
 	return SortSubsets(final)
 }
 
-type addressSet map[*api.EndpointAddress]struct{}
+// The sets of hosts must be de-duped, using IP+UID as the key.
+type addressKey struct {
```

Comment:
It's unique ports here, but when we divvy by subsets we allocate by unique
host ready status.  But this we originally Tims code and I may be missing a
case that could be simplfied

On Sep 15, 2015, at 10:28 PM, Abhi Shah <notifications@github.com> wrote:

In pkg/api/endpoints/util.go
<https://github.com/kubernetes/kubernetes/pull/13778#discussion_r39588390>:

>  	}
>
>  	// Finally, sort it.
>  	return SortSubsets(final)
>  }
>
> -type addressSet map[*api.EndpointAddress]struct{}
> +// The sets of hosts must be de-duped, using IP+UID as the key.
> +type addressKey struct {

I think we should just get rid of addressKey and use the endpointAddress as
the key itself.
We are inconsistent in the code here. we use addressKey in the map, but not
during hash generation of the address set.
We should be consistent.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13778/files#r39588390>.